### ISSUE TITLE
Why this script caused stack overflow exception?

### ISSUE THREAD

**Issue by ylxie** [pytorch/pytorch#7198]
## Issue description

I got a stack overflow exception for the following codes. Any help will be appreciated!!!

The exception happens when calling "loss.item()".

## Code example
```
import torch
import torch.autograd

import random

random.seed(1)
torch.manual_seed(1)

class DataGenerator:
    def __init__(self, feature_dim, max_seq_len, num_train_samples):
        self.feature_dim = feature_dim
        self.max_seq_len = max_seq_len
        self.num_train_samples = num_train_samples
    
    def get_next_batch(self, batch_size=1, num_batches=10):                
        for b in range(num_batches):
            samples = []
            for i in range(batch_size):
                length = random.randint(self.max_seq_len - 100, self.max_seq_len)
                features = [self.create_feature() for j in range(length)]
                target = random.randint(0, 1)
                samples.append((features, target))
            yield self.create_padded_batch(samples)
    
    def create_feature(self):
        tensor = torch.zeros(1, self.feature_dim)
        d = random.randint(0, self.feature_dim - 1)
        tensor[0][d] = 1.0
        return tensor

    def create_padded_batch(self, samples):
        sorted_samples = sorted(samples, key=lambda x:len(x[0]), reverse=True)
        inputs = [torch.cat(x[0], dim=0) for x in sorted_samples]
        input_lengths = [len(x) for x in inputs]
        targets = [x[1] for x in sorted_samples]
        target_batch = torch.FloatTensor(targets).view(-1,1)
        packed_input = torch.nn.utils.rnn.pad_sequence(inputs, batch_first=True)
        return packed_input, input_lengths, target_batch


class LSTMRNNModule(torch.nn.Module):
    def __init__(self, input_size, hidden_size):
        super(LSTMRNNModule, self).__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.lstm = torch.nn.LSTM(input_size, hidden_size, num_layers=1, bidirectional=False, batch_first=True, dropout=0)
        self.output_layer = torch.nn.Linear(hidden_size, 1)
    
    def forward(self, input, input_lengths):
        packed_input = torch.nn.utils.rnn.pack_padded_sequence(input, input_lengths, batch_first=True)
        out, (hidden, cstate) = self.lstm(packed_input)
        output = self.output_layer(hidden)
        return output

class Trainer(object):

    def __init__(self, feature_dim, hidden_size, data_set):
        self.data_set = data_set
        self.rnn = LSTMRNNModule(feature_dim, hidden_size)
        self.criterion = torch.nn.BCEWithLogitsLoss(size_average=True, reduce=True)
        self.optimizer = torch.optim.SGD(self.rnn.parameters(), lr=0.001, momentum=0.9)
    
    def train(self, max_iter, eval_every, batch_size, num_train_samples):

        for it in range(max_iter):            
            batch_num = 0
            for X,L,Y in self.data_set.get_next_batch(batch_size, num_train_samples):
                batch_num += 1
                loss = self.train_iter(X,L,Y)                
                print("{3}/{0}/{1}, Loss: {2:0.3f}".format(it + 1, max_iter, loss, batch_num))
    
    def train_iter(self, input, input_lengths, target):
        self.optimizer.zero_grad()
        output = self.rnn(input, input_lengths)
        output = output[0]
        loss = self.criterion(output, target)
        loss.backward()
        self.optimizer.step()
        return loss.item()


FEATURE_DIM = 1000
HIDDEN_SIZE = 32
MAX_SEQ_LEN = 1500
NUM_TRAIN_SAMPLES = 2000
MAX_ITER = 30
EVAL_EVERY = 1
BATCH_SIZE = 1

data_set = DataGenerator(FEATURE_DIM, MAX_SEQ_LEN, NUM_TRAIN_SAMPLES)

trainer = Trainer(FEATURE_DIM, HIDDEN_SIZE, data_set)

trainer.train(MAX_ITER, EVAL_EVERY, BATCH_SIZE, NUM_TRAIN_SAMPLES)

```

## System Info
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.60
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.13.1)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.0)
[conda] pytorch                   0.4.0           py36_cuda80_cudnn7he774522_1    pytorch
[conda] torchvision               0.2.0                     <pip>

**Comment by zou3519** [comment]
This is probably fixed by https://github.com/pytorch/pytorch/pull/6873 but I haven't verified the script doesn't stack overflow on a windows machine yet.

**Comment by SsnL** [comment]
Can you post the actual error message? Is it a SO error in Python?

**Comment by ylxie** [comment]
@zou3519 I haven't tried it on Linux yet. Will give it a try.

@SsnL, there is no error message, the python just crashes. I saw the stack overflow exception using both faulthandler and windbg.

Can you guys repro it using my scripts?

**Comment by zou3519** [comment]
@ylxie I'm unable to reproduce on Linux. I haven't tried on a Windows machine yet (will get to that soon).

If it is the same problem that #6873 fixes,  then it is possible that you'll see the stack overflow on Windows but not Linux. In addition, when you'll see the stack overflow on Windows is highly variable. Furthermore, python will just crash (on windows) without an error message.

**Comment by peterjc123** [comment]
@SsnL I've tried this using the binary build, So I think this problem is not the same as #6873, but may be related to a [problem](https://discuss.pytorch.org/t/rookie-not-getting-the-lstm-sine-time-sequence-example-to-work/16840) reported in the forum. 
@ylxie Could you please do the following things to locate the problem?
Change the flag `/Z7` in `setup.py` and `tools/build_pytorch_libs.bat` into `/ZI` and build again. And then, try to capture the error using some JIT debuggers like Visual Studio.

**Comment by ylxie** [comment]
@peterjc123 , after replacing /Z7 to /ZI I got error, '/ZI and /GL is incompatible', then I appened '/GL-'. but got the following error:

Module.cpp
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\functions.h(969): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\functions.h(973): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\tensoroperators.h(47): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\tensoroperators.h(53): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\tensoroperators.h(56): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\lib\tmp_install\include\aten\expandutils.h(19): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\csrc\autograd\variable.h(462): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion
c:\work\test\pytorch\pytorch\torch\csrc\autograd\variable.h(471): error C2397: conversion from 'long' to 'uint32_t' requires a narrowing conversion

**Comment by peterjc123** [comment]
@ylxie Sorry for the typo. It should be `/Zi`.

**Comment by ylxie** [comment]
@peterjc123 there is still errors:

C:\Work\Test\pytorch\pytorch\torch/csrc/utils/hash.h(78): error C2668: 'torch::_hash_detail::dispatch_hash': ambiguous call to overloaded function

**Comment by peterjc123** [comment]
@ylxie It should not be a question if you use MSVC 14.11. How did you try to build it?

**Comment by ylxie** [comment]
@peterjc123 @zou3519 @SsnL, Update on you guys that: the issue seems gone in the latest version (up to commit: 0859f0e3e6ceb02f56551271476009ad7300e697). I cannot repo it using my locally built binaries. Looks like an issue in the 0.4.0 version only.

Btw, the version # of my locally built is: 0.5.0a0+0859f0e.

Looking forward to the official pytorch 1.0 release. Cheers!
